### PR TITLE
[ATen] Add reduction tag to missing base reduction ops (re-land of #188902)

### PR DESCRIPTION
**Pull request by YAshhh29**
Re-land of #188902 with the fix for the MPS `mode` OpInfo interaction that caused the original PR to be auto-reverted.

### What was reverted

#188902 added the `reduction` tag to 14 `native_functions.yaml` entries. That caused `test/test_ops.py::test_reduction_ops_reduce` to be generated for the `mode` op for the first time. The new test `test_reduction_ops_reduce_mode_mps` failed with `Unexpected success` on trunk, triggering the auto-revert.

### Root cause

1. `test_reduction_ops_reduce` is parametrized over ops with `torch.Tag.reduction`
2. Before #188902, `mode` had no tag, so the test was never generated for it
3. After the tag was added, `test_reduction_ops_reduce_mode_mps` is created
4. `mode`'s OpInfo carries a blanket `TestCommon` `expectedFailure` on MPS, because `aten::mode` dispatches to `CPU, CUDA, XPU` only
5. The test's per-sample loop does `if "dim" not in sample.kwargs: continue`, and `sample_inputs_mode` passed `dim` **positionally** — so every sample was skipped and `aten::mode` was never invoked
6. Trivial pass + `expectedFailure` = `Unexpected success` = red trunk

Nothing was wrong with the tag additions themselves. The failure came entirely from the interaction between the `mode` OpInfo's blanket decorator and the newly-generated test.

### Fix

`sample_inputs_mode` now passes `dim` and `keepdim` as **kwargs** instead of positionally. No new decorators are added and no existing ones are touched.

With this change `test_reduction_ops_reduce` actually invokes `aten::mode`:

- On CPU/CUDA/XPU, `torch.mode(x, dim=k)` returns a `(values, indices)` namedtuple. The test's `isinstance(result, torch.Tensor)` guard skips the numel assertion for those samples, as intended.
- On MPS, `aten::mode` raises `NotImplementedError`, which the existing blanket `expectedFailure` catches as an expected failure.
- If MPS support is added later, that xfail will fire `Unexpected success` and prompt its removal — preserving the maintenance signal @Skylion007 asked for, rather than silencing it with a skip.

Fixing this at the sample-inputs level addresses the cause rather than suppressing the symptom.

### Newly tagged ops

- `nanmean`, `nanmean.out` (mirrors `mean.*`)
- `median`, `median.dim`, `median.dim_values`
- `nanmedian`, `nanmedian.dim`, `nanmedian.dim_values` (mirror `min.dim` / `max.dim`)
- `mode`, `mode.values` (dim-reduced like `min.dim`)
- `trace` (sum over diagonal -> scalar)
- `dist` (`norm(a - b)` -> scalar)
- `special_logsumexp`, `special_logsumexp.out` (mirrors `logsumexp.*`)

### Audit of the other tagged ops

Confirmed none share the MPS blanket-xfail interaction that `mode` had:

| Op | OpInfo state on MPS |
| --- | --- |
| `nanmean` | No blanket xfail |
| `median` | `dtypesIfMPS` set -> supported |
| `nanmedian` | `dtypesIfMPS` set -> supported |
| `trace` | `dtypesIfMPS` set -> supported |
| `dist` | No blanket xfail |
| `special_logsumexp` (via `logsumexp`) | No blanket xfail |

### Diff summary

- `aten/src/ATen/native/native_functions.yaml`: +14 (unchanged from #188902)
- `torch/testing/_internal/common_methods_invocations.py`: +16 / -12 (`sample_inputs_mode`)

Total: 2 files, +30 / -12.

### Context

Follow-up to #129020, per @eellison's invitation on #169944 ("if there are any missing ops, please submit a pr to add them").

cc @eellison @drisspg

### CODE REVIEW

**Review by eellison — APPROVED** [review]

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -18816,6 +18816,11 @@ def sample_inputs_abs(op_info, device, dtype, requires_grad, op_kwargs=None, **k
                    dtypes=(torch.float32,),
                    device_type="cuda",
                ),
+               # sample_inputs_mode passes `dim` positionally, so test_reduction_ops_reduce
+               # skips every sample and trivially passes without invoking aten::mode; the
+               # blanket MPS expectedFailure below then flags it as an unexpected success.
+               DecorateInfo(unittest.skip("mode not implemented on MPS"),
```
Comment (Skylion007) [inline torch/testing/_internal/common_methods_invocations.py]: Wait, this should be UnitTest.expectedSuccess or xfail here, not skip please in case we fix it later

**Review by Skylion007 — DISMISSED** [review]

**Comment by YAshhh29** [comment]
Thanks for flagging — one wrinkle with using `expectedFailure` at that OpInfo position: the test currently passes *trivially* on MPS (the sample loop skips every sample because `dim` isn't in `sample.kwargs`), so `expectedFailure` would immediately fire "Unexpected success" — the exact failure this PR is trying to fix.

To address the "we should notice when this gets fixed" concern, I've pushed a different approach that removes the OpInfo change entirely and instead updates `sample_inputs_mode` to pass `dim`/`keepdim` as kwargs. Now:

- `test_reduction_ops_reduce` actually invokes `aten::mode` on MPS (rather than skipping every sample)
- MPS raises `NotImplementedError`, which the existing blanket `TestCommon` `expectedFailure` on the `mode` OpInfo catches as expected
- When MPS support is added later, that xfail will fire "Unexpected success" and prompt its removal — the maintenance signal you were pointing at

Diff is now `sample_inputs_mode` (+16/-17) plus the 14 `native_functions.yaml` tag additions. No new decorators. PTAL.

**Comment by YAshhh29** [comment]
Gentle bump — this is waiting on two maintainer-only actions:

1. CI for the amended commit `a37286f` hasn't been triggered yet (4 workflows awaiting approval).
2. @Skylion007's `Changes requested` review still shows as blocking, but the specific line it was on has been replaced. The new approach fixes `sample_inputs_mode` (rather than adding a skip decorator) so the maintenance-signal concern is addressed at its root — details in the reply thread above.

@Skylion007 could you take another look? @eellison / @zou3519 also happy for either of you to weigh in. Thanks!

**Comment by YAshhh29** [comment]
CI complete. The two remaining failures are on the active MPS-runner SEV #190674 ("Unrelated MacOS MPS test jobs time out (30 min) due to stale cpp_extension lock file poisoning self-hosted runners"):

- `macos-m1-14`: timed out after 126m
- `macos-m2-15`: timed out after 120m
- `macos-m2-26` (same job, sibling shard): passed clean in 48m

Failing test in both timeouts: `test/test_mps.py::TestBinaryIteratorConformance::test_simple_add_bfloat16_float32_float32_shape2`. That test uses `torch.utils.cpp_extension.load_inline` to compile a Metal extension for element-wise `torch.add` bfloat16/float32 — i.e. it hits the exact `cpp_extension` lock-file path named in the SEV. It doesn't touch reduction tags, OpInfos, `sample_inputs_mode`, or any of the ops changed here.

Consistent with the same shard timing out on the original #188902 (which had a completely different fix), so the failure is runner-side, not code-side. Dr.CI also marks both as `AI verdict: not related`.

@eellison this looks ready for `@pytorchbot merge -i` given the SEV. Thanks!

**Comment by YAshhh29** [comment]
Rebased onto current main (`7c69012`) — the branch was 1967 commits behind, with merge base `e6e0fe7` from mid-July.

Rationale: the two MPS timeouts on the previous run were the stale `FileBaton` lock issue from #190674. That issue is now closed, and malfet's stated remediation there is "rebase past a84391f so the tests stop leaving locks behind." `a84391f` was not in this branch and is now included, so this should get a clean MPS run rather than needing a force-merge.

The diff is unchanged from what @eellison approved — same 2 files, `+30/-12`, rebase applied with no conflicts. I also re-verified the three assumptions the fix rests on against today's main:

- `mode`'s OpInfo still carries the blanket `DecorateInfo(unittest.expectedFailure, 'TestCommon', device_type='mps')`
- `test_reduction_ops_reduce` is still defined inside `TestCommon`, so that xfail covers it
- `aten::mode` still dispatches to `CPU, CUDA, XPU` only, so MPS raises `NotImplementedError` and the xfail catches it as expected

Could a maintainer approve the workflows? 4 are awaiting approval. @eellison @Skylion007 @zou3519 — thanks!
